### ISSUE TITLE
Implement PPO actor-critic network

### ISSUE THREAD

**Issue by tgrunnagle** [tgrunnagle/go-towerfall#39]
## Summary

Implement the neural network architecture for the PPO (Proximal Policy Optimization) actor-critic agent. This includes the policy network (actor) that outputs action probabilities and the value network (critic) that estimates state values for advantage estimation.

**Task ID:** TASK-021
**Parent Task:** TASK-P006 (PPO Agent Implementation)
**Epic:** #14

## Dependencies

- **#25 (TASK-007)**: Configure bot2 project dependencies - Provides PyTorch and NumPy
- **#30 (TASK-011)**: Define discrete action space - Provides `ACTION_SPACE_SIZE = 27` discrete actions
- **#35 (TASK-012)**: Implement observation space normalization - Provides observation vector size (114 values by default)

## Background

From the intake requirements and research:
- PPO is the chosen RL algorithm for training ML bots
- The observation space is a normalized vector of 114 values (own player: 14, other players: 3x12=36, arrows: 8x8=64)
- The action space has 27 discrete actions (8 movement, 16 aim directions, 2 shoot, 1 no-op)
- All observations are normalized to [-1, 1] range for neural network compatibility

#### 1.1 Shared Feature Extractor
A shared MLP backbone that processes the observation vector:

```
Input (114) -> Linear(114, 256) -> ReLU -> Linear(256, 256) -> ReLU -> features (256)
```

#### 1.2 Actor (Policy) Head
Outputs action logits for the discrete action space:

```
features (256) -> Linear(256, 128) -> ReLU -> Linear(128, 27) -> action_logits
```

The action distribution is a Categorical distribution over 27 actions.

#### 1.3 Critic (Value) Head
Outputs a scalar state value estimate:

```
features (256) -> Linear(256, 128) -> ReLU -> Linear(128, 1) -> value
```

#### 2.1 File Structure
```
bot2/
  agent/
    __init__.py
    network.py          # Neural network architecture
    actor_critic.py     # ActorCritic wrapper class
    distributions.py    # Action distribution utilities
  tests/
    test_network.py     # Unit tests
```

#### 2.2 Core Classes

**ActorCriticNetwork class:**

```python
import torch
import torch.nn as nn
from torch.distributions import Categorical

class ActorCriticNetwork(nn.Module):
    """PPO Actor-Critic neural network with shared feature extractor."""

    def __init__(
        self,
        observation_size: int = 114,
        action_size: int = 27,
        hidden_size: int = 256,
        actor_hidden: int = 128,
        critic_hidden: int = 128,
    ):
        super().__init__()

        # Shared feature extractor
        self.features = nn.Sequential(
            nn.Linear(observation_size, hidden_size),
            nn.ReLU(),
            nn.Linear(hidden_size, hidden_size),
            nn.ReLU(),
        )

        # Actor head (policy)
        self.actor = nn.Sequential(
            nn.Linear(hidden_size, actor_hidden),
            nn.ReLU(),
            nn.Linear(actor_hidden, action_size),
        )

        # Critic head (value function)
        self.critic = nn.Sequential(
            nn.Linear(hidden_size, critic_hidden),
            nn.ReLU(),
            nn.Linear(critic_hidden, 1),
        )

        # Initialize weights
        self._initialize_weights()

    def _initialize_weights(self):
        """Initialize network weights using orthogonal initialization."""
        for module in self.modules():
            if isinstance(module, nn.Linear):
                nn.init.orthogonal_(module.weight, gain=nn.init.calculate_gain('relu'))
                nn.init.zeros_(module.bias)

        # Smaller initialization for output layers
        nn.init.orthogonal_(self.actor[-1].weight, gain=0.01)
        nn.init.orthogonal_(self.critic[-1].weight, gain=1.0)

    def forward(self, obs: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
        """Forward pass returning action logits and state value."""
        features = self.features(obs)
        action_logits = self.actor(features)
        value = self.critic(features)
        return action_logits, value.squeeze(-1)

    def get_action_and_value(
        self,
        obs: torch.Tensor,
        action: torch.Tensor | None = None,
        deterministic: bool = False,
    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
        """
        Get action, log probability, entropy, and value.

        Args:
            obs: Observation tensor of shape (batch, obs_size)
            action: Optional action tensor. If None, sample from policy.
            deterministic: If True and action is None, use argmax instead of sampling.

        Returns:
            action: Selected action
            log_prob: Log probability of the action
            entropy: Entropy of the action distribution
            value: Estimated state value
        """
        action_logits, value = self.forward(obs)
        dist = Categorical(logits=action_logits)

        if action is None:
            if deterministic:
                action = action_logits.argmax(dim=-1)
            else:
                action = dist.sample()

        log_prob = dist.log_prob(action)
        entropy = dist.entropy()

        return action, log_prob, entropy, value

    def get_value(self, obs: torch.Tensor) -> torch.Tensor:
        """Get state value only (used during rollout for bootstrapping)."""
        features = self.features(obs)
        return self.critic(features).squeeze(-1)
```

#### 2.3 Design Decisions

1. **Shared Feature Extractor**: Reduces parameters and allows actor and critic to share learned representations. Common in PPO implementations.

2. **Orthogonal Initialization**: Standard for PPO networks, helps with training stability.

3. **Hidden Sizes**: 256 for shared layers, 128 for heads. Reasonable starting point for a 114-dimensional input space. Can be tuned via hyperparameters (TASK-024).

4. **Categorical Distribution**: Uses PyTorch's built-in Categorical for proper log probability and entropy calculations.

5. **Separate get_value Method**: Useful during rollout collection when only the value is needed for bootstrapping.

### 3. Integration Points

- **Observation Size**: Imported from `bot2.observation.observation_space.ObservationConfig.total_size`
- **Action Size**: Imported from `bot2.actions.ACTION_SPACE_SIZE`
- **Hyperparameters**: Network dimensions should be configurable via the config system (TASK-024)

### 4. Device Support

The network should support both CPU and CUDA:

```python
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
network = ActorCriticNetwork().to(device)
```

## Acceptance Criteria

- [ ] `ActorCriticNetwork` class implemented in `bot2/agent/network.py`
- [ ] Network accepts observation tensor of shape `(batch, 114)` and outputs action logits `(batch, 27)` and value `(batch,)`
- [ ] `get_action_and_value()` method returns action, log_prob, entropy, and value
- [ ] `get_value()` method returns value only for efficient bootstrapping
- [ ] Orthogonal weight initialization applied correctly
- [ ] Network supports both CPU and CUDA devices
- [ ] Unit tests verify:
  - [ ] Forward pass produces correct output shapes
  - [ ] `get_action_and_value()` returns valid log probabilities (negative values)
  - [ ] `get_action_and_value()` with provided action returns correct log_prob for that action
  - [ ] Deterministic mode returns argmax action
  - [ ] Entropy values are non-negative
  - [ ] Network can be moved to CUDA (if available)
- [ ] Type hints using native Python types per CLAUDE.md guidelines
- [ ] Imports relative to bot2 directory

## Test Cases

```python
def test_forward_shapes():
    network = ActorCriticNetwork(observation_size=114, action_size=27)
    obs = torch.randn(32, 114)  # Batch of 32
    action_logits, value = network(obs)
    assert action_logits.shape == (32, 27)
    assert value.shape == (32,)

def test_get_action_and_value():
    network = ActorCriticNetwork()
    obs = torch.randn(16, 114)
    action, log_prob, entropy, value = network.get_action_and_value(obs)
    assert action.shape == (16,)
    assert log_prob.shape == (16,)
    assert entropy.shape == (16,)
    assert value.shape == (16,)
    assert (action >= 0).all() and (action < 27).all()
    assert (log_prob <= 0).all()  # Log probs are negative
    assert (entropy >= 0).all()  # Entropy is non-negative

def test_deterministic_action():
    network = ActorCriticNetwork()
    obs = torch.randn(1, 114)
    action1, _, _, _ = network.get_action_and_value(obs, deterministic=True)
    action2, _, _, _ = network.get_action_and_value(obs, deterministic=True)
    assert action1 == action2  # Deterministic should be consistent

def test_provided_action_log_prob():
    network = ActorCriticNetwork()
    obs = torch.randn(8, 114)
    actions = torch.randint(0, 27, (8,))
    _, log_prob, _, _ = network.get_action_and_value(obs, action=actions)
    assert log_prob.shape == (8,)
```

## Notes

- The hyperparameter configuration (TASK-024) will add ability to customize hidden layer sizes
- TASK-022 (PPO training loop) will use this network for rollout collection and policy updates
- TASK-023 (model serialization) will handle saving/loading trained network weights
- Consider adding layer normalization if training is unstable (can be added in future iteration)

## References

- PPO Paper: https://arxiv.org/abs/1707.06347
- CleanRL PPO Implementation: https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo.py
- Stable-Baselines3 PPO: https://github.com/DLR-RM/stable-baselines3/tree/master/stable_baselines3/ppo
